### PR TITLE
feat: add GMP native token handler adapter

### LINKED ISSUE #231: Adapter for native currency transfers
We want to support transfers in native currency (e.g. EVM networks - ETH). There is already an [adapter](https://github.com/sygmaprotocol/native-currency-adapter) written by the Solidity team that could be re-used.
(We should talk with them for more context about this adapter)

## Implementation details
Re-use the adapter that already exists for the new handler.

## Testing details
passing unit and e2e tests

## Acceptance Criteria
We can bridge native currency as ETH.

### PR DESCRIPTION
**Pull request by nmlinaric**
## Description
Adds adapter for native token trasnfers via GMP.

## Related Issue Or Context

Closes: #231 

## How Has This Been Tested? Testing details.
unit and e2e tests 

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation

## Checklist:
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have ensured that all acceptance criteria (or expected behavior) from issue are met
- [ ] I have updated the documentation locally and in chainbridge-docs.
- [ ] I have added tests to cover my changes.
- [ ] I have ensured that all the checks are passing and green, I've signed the CLA bot

### CODE REVIEW

**Comment by nmlinaric** [comment]
I still need to add tests for deposit logic and some edge cases - I'll do it next week

File: contracts/adapters/nativeTokens/DestinationAdapter.sol [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    }
+
+    function transferFunds(address originAdapter, bytes calldata data) external {
+        if(_originAdapter != originAdapter) revert InvalidOriginAdapter(originAdapter);
+        if(msg.sender != _gmpAddress) revert InvalidHandler(msg.sender);
+
+        uint256 amount;
+        address recipient;
+
+        (recipient, amount) = abi.decode(data, (address, uint256));
```
Comment (lastperson) [inline contracts/adapters/nativeTokens/DestinationAdapter.sol:58]:
```suggestion
    function transferFunds(address originAdapter, address payable recipient, uint256 amount) external {
        if(_originAdapter != originAdapter) revert InvalidOriginAdapter(originAdapter);
        if(msg.sender != _gmpAddress) revert InvalidHandler(msg.sender);
```

File: contracts/adapters/nativeTokens/DestinationAdapter.sol [hunk 2]
```diff
@@ -0,0 +1,64 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity ^0.8.11;
+
+import "@openzeppelin/contracts/access/AccessControl.sol";
+
+contract DestinationAdapter is AccessControl {
+    event FundsTransferred(address recipient, uint256 amount);
+
+    address _originAdapter;
+    address _gmpAddress;
```
Comment (lastperson) [inline contracts/adapters/nativeTokens/DestinationAdapter.sol:10]:
```suggestion
    address immutable _gmpAddress;
```

**Review by lastperson — CHANGES_REQUESTED** [review]
If native currencies are of the same value between the origin and destination, then perhaps we should have a single adapter implementation deployed on both chains to facilitate movement in both directions?

File: contracts/adapters/interfaces/IDepositAdapterTarget.sol [hunk 3]
```diff
@@ -0,0 +1,6 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity ^0.8.11;
+
+interface IDepositAdapterTarget {
+    function transferFunds(address originAdapter, bytes calldata depositData) external;
```
Comment (lastperson) [inline contracts/adapters/interfaces/IDepositAdapterTarget.sol:5]:
```suggestion
    function transferFunds(address originAdapter, address payable recipient, uint256 amount) external;
```

File: contracts/adapters/nativeTokens/DestinationAdapter.sol [hunk 4]
```diff
@@ -0,0 +1,40 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity ^0.8.11;
+
+import "@openzeppelin/contracts/access/AccessControl.sol";
+
+contract DestinationAdapter is AccessControl {
```
Comment (lastperson) [inline contracts/adapters/nativeTokens/DestinationAdapter.sol:6]:
```suggestion
contract DestinationAdapter is IDepositAdapterTarget, AccessControl {
```
To make sure the transferFunds is compatible.

File: contracts/adapters/nativeTokens/OriginAdapter.sol [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+    modifier onlyAdmin() {
+        if (!hasRole(DEFAULT_ADMIN_ROLE, msg.sender)) revert SenderNotAdmin();
+        _;
+    }
+
+    constructor(IBridge bridge, IBasicFeeHandler feeHandler, bytes32 resourceID) {
+        _bridge = bridge;
+        _resourceID = resourceID;
+        _feeHandler = feeHandler;
```
Comment (lastperson) [inline contracts/adapters/nativeTokens/OriginAdapter.sol:31]: Any reason why the fee handler is set separately instead of being fetched from the bridge when the fee calculation is required?

**Review by lastperson — CHANGES_REQUESTED** [review]

**Review by lastperson — APPROVED** [review]

**Review by mpetrun5 — COMMENTED** [review]
The implementation is OK, though I would consider adapter implementations be part of another repo with them. 
`sygma-solidity` IMO should only be for core solidity stuff.

**Comment by nmlinaric** [comment]
Closing this as for the time being it will not be used in production
